### PR TITLE
ESQL: Track `Block`s from topn

### PR DESCRIPTION
**Pull request by nik9000**
This creates tracked `Block`s from topn and closes the blocks passed to it.

### CODE REVIEW

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java [hunk 1]
```diff
@@ -80,15 +83,26 @@ public final void testSimpleLargeInput() {
...
+            CircuitBreaker breaker = bigArrays.breakerService().getBreaker(CircuitBreaker.REQUEST);
+            BlockFactory blockFactory = BlockFactory.getInstance(breaker, bigArrays);
+            Exception e = expectThrows(
+                CircuitBreakingException.class,
+                () -> drive(simple(bigArrays).get(new DriverContext(bigArrays, blockFactory)), input.iterator())
+            );
+            assertThat(e.getMessage(), equalTo(MockBigArrays.ERROR_MESSAGE));
+            assertThat(bigArrays.breakerService().getBreaker(CircuitBreaker.REQUEST).getUsed(), equalTo(0L));
+        } finally {
+            Releasables.close(() -> Iterators.map(input.iterator(), p -> () -> p.releaseBlocks()));
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:104]: Boom! my head exploded!! ;-) So this is inferred to `Iterators.<Page, Releasable>map(..`. Maybe just use a method reference here to avoid the extra lambda arrows ( as suggested by your favourite IDE! ) 
```
Iterators.map(input.iterator(), p -> p::releaseBlocks)
```

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java [hunk 2]
```diff
@@ -80,15 +83,26 @@ public final void testSimpleLargeInput() {
...
-            () -> assertSimple(new DriverContext(bigArrays, blockFactory), between(1_000, 10_000))
-        );
-        assertThat(e.getMessage(), equalTo(MockBigArrays.ERROR_MESSAGE));
-        assertThat(bigArrays.breakerService().getBreaker(CircuitBreaker.REQUEST).getUsed(), equalTo(0L));
+        /*
+         * We build two CircuitBreakers - one for the input blocks and one for the operation itself.
+         * The input blocks don't count against the memory usage for the limited operator that we
+         * build.
+         */
+        List<Page> input = CannedSourceOperator.collectPages(simpleInput(driverContext().blockFactory(), between(1_000, 10_000)));
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:91]: We want this blockFactory to never circuit break, right? Just do the accountancy.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java [hunk 3]
```diff
@@ -139,10 +153,11 @@ protected final List<Page> oneDriverPerPageList(Iterator<List<Page>> source, Sup
     }
 
     private void assertSimple(DriverContext context, int size) {
-        List<Page> input = CannedSourceOperator.collectPages(simpleInput(size));
+        List<Page> input = CannedSourceOperator.collectPages(simpleInput(context.blockFactory(), size));
+        List<Page> inputClone = CannedSourceOperator.deepCopyOf(input);
```
Comment (ChrisHegarty) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:157]: why do we clone here?

**Review by ChrisHegarty — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:91]: Yeah. Right now we actually don't assert that we release all the blocks either. I'll grab that in a follow up. The trouble, I think, is just a missing `try` block around processing the page. I *think* it's that simple, but I want to dig more into the failure cases in a follow up.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java (hunk 1 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:104]: That kept not working. Let me try again.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java (hunk 3 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java:157]: Adding a comment:
```
// Clone the input so that the operator can close it, then, later, we can read it again to build the assertion.
```
